File: mac/Sources/SoundboyCore/AnalysisCache.swift
import Foundation

/// Remembers BPM/key results per file (keyed by path + size + modified time) in
/// ~/Library/Application Support/SOUNDBOY/analysis.json so each file is analyzed only once.
public final class AnalysisCache: @unchecked Sendable {
    private var entries: [String: AnalysisResult]
    private let lock = NSLock()
    private let file: URL

    public init(directory: URL = AppPaths.support) {
        file = directory.appendingPathComponent("analysis.json")
        entries = (try? JSONDecoder().decode([String: AnalysisResult].self, from: Data(contentsOf: file))) ?? [:]
    }

    static func key(for url: URL) -> String? {
        guard let a = try? FileManager.default.attributesOfItem(atPath: url.path),
              let size = a[.size] as? NSNumber, let date = a[.modificationDate] as? Date else { return nil }
        return "\(url.standardizedFileURL.path.lowercased())|\(size.int64Value)|\(Int64(date.timeIntervalSince1970 * 1000))"
    }

    public func get(_ url: URL) -> AnalysisResult? {
        guard let k = Self.key(for: url) else { return nil }
        lock.lock(); defer { lock.unlock() }
        return entries[k]
    }

    public func put(_ url: URL, _ r: AnalysisResult) {
        guard let k = Self.key(for: url) else { return }
        lock.lock(); defer { lock.unlock() }
        entries[k] = r
    }

    public func save() {
        lock.lock()
        let data = try? JSONEncoder().encode(entries)
        lock.unlock()
        try? data?.write(to: file, options: .atomic)
    }
}

File: SOUNDBOY/Model/AnalysisCache.cs
using System.Text.Json;
using SOUNDBOY.Audio;

namespace SOUNDBOY.Model;

/// <summary>
/// Remembers BPM/key results per file (keyed by path + size + modified time) in
/// %APPDATA%\SOUNDBOY\analysis.json so each file is analyzed only once.
/// </summary>
public sealed class AnalysisCache
{
    sealed record Entry(double? Bpm, string? Key, string? Camelot);

    static string FilePath => Path.Combine(AppSettings.Dir, "analysis.json");
    readonly Dictionary<string, Entry> entries;
    readonly object sync = new();

    public AnalysisCache()
    {
        try
        {
            entries = File.Exists(FilePath)
                ? JsonSerializer.Deserialize<Dictionary<string, Entry>>(File.ReadAllText(FilePath)) ?? new()
                : new();
        }
        catch
        {
            entries = new();
        }
    }

    static string? KeyFor(string path)
    {
        try
        {
            var fi = new FileInfo(path);
            return fi.Exists ? $"{fi.FullName.ToLowerInvariant()}|{fi.Length}|{fi.LastWriteTimeUtc.Ticks}" : null;
        }
        catch
        {
            return null;
        }
    }

    public AnalysisResult? Get(string path)
    {
        var k = KeyFor(path);
        lock (sync)
            return k != null && entries.TryGetValue(k, out var e) ? new AnalysisResult(e.Bpm, e.Key, e.Camelot) : null;
    }

    public void Put(string path, AnalysisResult r)
    {
        var k = KeyFor(path);
        if (k == null) return;
        lock (sync) entries[k] = new Entry(r.Bpm, r.Key, r.Camelot);
    }

    public void Save()
    {
        try
        {
            Directory.CreateDirectory(AppSettings.Dir);
            string json;
            lock (sync) json = JsonSerializer.Serialize(entries);
            File.WriteAllText(FilePath, json);
        }
        catch { }
    }
}
